File: problem_3/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func productArray(arr []int) []int {
	n := len(arr)
	prod_arr := []int{}
	for i := 0; i < n; i++ {
		prod_arr = append(prod_arr, 1)
	}

	temp := 1
	for i := 0; i < n; i++ {
		prod_arr[i] = temp
		temp *= arr[i]
	}

	temp = 1
	for i := n - 1; i >= 0; i-- {
		prod_arr[i] *= temp
		temp *= arr[i]
	}

	return prod_arr

}

func main() {
	file, err := os.Open("test_cases.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		str_arr := strings.Split(scanner.Text(), ",")
		arr := []int{}
		for _, i := range str_arr {
			j, err := strconv.Atoi(i)
			if err != nil {
				fmt.Println(err)
			}
			arr = append(arr, j)
		}

		fmt.Printf("Product array: %v\n", productArray(arr))

		if err := scanner.Err(); err != nil {
			fmt.Println(err)
		}
	}
}

File: problem_3/solution.py
"""
A Product Array Puzzle without division operator
"""

def product_array(arr):
    n = len(arr)
    prod_arr = [1] * n
    temp = 1
    for i in range(n):
        prod_arr[i] = temp
        temp *= arr[i]

    temp = 1
    for i in range(n-1, -1, -1):
        prod_arr[i] *= temp
        temp *= arr[i]

    return prod_arr


if __name__ == '__main__':
    filename = "test_cases.txt"
    with open(filename) as file:
        content = file.readline()
        while content:

            arr = [int(a) for a in content.split(',')]
            print("Product array: {}".format(product_array(arr)))

            content = file.readline()
